packages/evm/contracts/Integrity.sol

```diff
@@ -246,32 +248,56 @@ library Integrity {
             childrenBounds
         );
 
-        if (typeTree.paramType != ParameterType.AbiEncoded) {
+        if (typeTree.paramType != ParameterType.Calldata) {
             revert UnsuitableRootNode();
         }
     }
 
-    function compatiblechildTypeTree(
+    function _compatibleSiblingTypes(
         ConditionFlat[] memory conditions,
         uint256 index,
         Topology.Bounds[] memory childrenBounds
     ) private pure {
         uint256 start = childrenBounds[index].start;
         uint256 end = childrenBounds[index].end;
 
-        bytes32 id = typeTreeId(
-            Topology.typeTree(conditions, start, childrenBounds)
-        );
         for (uint256 j = start + 1; j < end; ++j) {
             if (
-                id !=
-                typeTreeId(Topology.typeTree(conditions, j, childrenBounds))
+                !_isTypeMatch(conditions, start, j, childrenBounds) &&
+                !_isTypeEquivalent(conditions, start, j, childrenBounds)
             ) {
                 revert UnsuitableChildTypeTree(index);
             }
         }
     }
 
+    function _isTypeMatch(
+        ConditionFlat[] memory conditions,
+        uint256 i,
+        uint256 j,
+        Topology.Bounds[] memory childrenBounds
+    ) private pure returns (bool) {
+        return
+            typeTreeId(Topology.typeTree(conditions, i, childrenBounds)) ==
+            typeTreeId(Topology.typeTree(conditions, j, childrenBounds));
+    }
+
+    function _isTypeEquivalent(
+        ConditionFlat[] memory conditions,
+        uint256 i,
+        uint256 j,
+        Topology.Bounds[] memory childrenBounds
+    ) private pure returns (bool) {
+        ParameterType leftParamType = Topology
+            .typeTree(conditions, i, childrenBounds)
+            .paramType;
+        return
+            (leftParamType == ParameterType.Calldata ||
```

Comment:
Does this mean we only allow siblings to occur in a order `AbiEncoded`, `Dynamic`, but not the reverse `Dynamic`, `AbiEncoded`?

Reply:
Yes. The decoder always uses first child to get TypeDefinition. Integrity ensures subsequent ones have to be something equivalent

Reply:
tricky for users to be aware of this detail. would be too expensive to scan all dynamic siblings in case one has children types?

Reply:
Prohibitively expensive to scan and search during execution. We just wanna get the first one.